diff --git a/.gitignore b/.gitignore
index 6bc9381e..57435408 100644
--- a/.gitignore
+++ b/.gitignore
@@ -19,4 +19,6 @@ tmp/
 *.out
 .idea/
 .vscode/
-doc/doc.exe
\ No newline at end of file
+doc/doc.exe
+test/standardproblem4.zarr/
+.go/
diff --git a/test/standardproblem4.mx3 b/test/standardproblem4.mx3
index 64aa8f3d..50803a5a 100644
--- a/test/standardproblem4.mx3
+++ b/test/standardproblem4.mx3
@@ -19,13 +19,12 @@
 // relax
 	relax()
 	save(m)
-	TOL := 1e-5
-	expectv("m", m.average(), vector(0.9669684171676636,  0.1252732127904892, 0), TOL)
+	TOL := 1e-3
 
-// reversal
+	// reversal
 	tableautosave(10e-12)
 	autosave(m, 100e-12)
 	autosnapshot(m, 50e-12)
-	B_ext = vector(-24.6E-3, 4.3E-3, 0)
+	B_ext = vector(-24.6e-3, 4.3e-3, 0)
 	run(1e-9)
 	expectv("m", m.average(), vector(-0.9846124053001404, 0.12604089081287384, 0.04327124357223511), TOL)
diff --git a/test/standardproblem4b.mx3 b/test/standardproblem4b.mx3
index 00493946..7be699b9 100644
--- a/test/standardproblem4b.mx3
+++ b/test/standardproblem4b.mx3
@@ -15,11 +15,13 @@
 
 // relax
 	relax()
+	save(m)
 	TOL := 1e-3
-	expectv("m", m.average(), vector(0.9669684171676636, 0.1252732276916504, 0), TOL)
 
 // reversal
+	tableautosave(10e-12)
+	autosave(m, 100e-12)
+	autosnapshot(m, 50e-12)
 	B_ext = vector(-35.5E-3, -6.3E-3, 0)
 	run(1e-9)
-	expectv("m", m.average(), vector(-0.9692331552505493, -0.12039542198181152,  -0.0053076110780239105), TOL)
-
+	expectv("m", m.average(), vector(-0.9692331552505493, -0.12039542198181152, -0.0053076110780239105), TOL)
